calico_cni_k8s_test.go

```diff
@@ -885,6 +885,55 @@ var _ = Describe("CalicoCni", func() {
 					_, err = DeleteContainer(netconf, netnspath, "")
 					Expect(err).ShouldNot(HaveOccurred())
 				})
+
+				Context("when IPAM returns an empty result", func() {
+					It("should try to release any IP addresses", func() {
+						netConf := fmt.Sprintf(`
+				{
+			      "cniVersion": "%s",
+				  "name": "net2",
+				  "type": "calico",
+				  "etcd_endpoints": "http://%s:2379",
+			 	  "ipam": {
+			    	 "type": "echo-ipam",
+				 "return": "{}"
+			         },
+					"kubernetes": {
+					  "k8s_api_root": "http://127.0.0.1:8080"
+					 },
+					"policy": {"type": "k8s"},
+					"log_level":"info"
+				}`, cniVersion, os.Getenv("ETCD_IP"))
+
+						// Now create a K8s pod
+						config, err := clientcmd.DefaultClientConfig.ClientConfig()
+						Expect(err).NotTo(HaveOccurred())
+						clientset, err := kubernetes.NewForConfig(config)
+						Expect(err).NotTo(HaveOccurred())
+
+						name := fmt.Sprintf("run%d-pool", rand.Uint32())
+						pod, err := clientset.Pods(K8S_TEST_NS).Create(&v1.Pod{
+							ObjectMeta: v1.ObjectMeta{
+								Name: name,
+							},
+							Spec: v1.PodSpec{Containers: []v1.Container{{
+								Name:  fmt.Sprintf("container-%s", name),
+								Image: "ignore",
+							}}},
+						})
+						Expect(err).NotTo(HaveOccurred())
+
+						logger.Infof("Created POD object: %v", pod)
+
+						_, netnspath, _, _, _, _, _, err := CreateContainer(netConf, name, "")
```

Comment:
So this seems to be working (can tell from the logs), but I don't know how to assert that the IPs have been actually released by the plugin.

I could update `echo-ipam` to do some fake IP allocation / releasing perhaps using files on disk and then assert that the file is removed in the test?

@tomdee